### ISSUE TITLE
Generic type inference prefers conditional types

### ISSUE THREAD

**Issue by bowenni** [microsoft/TypeScript#31766]
**TypeScript Version:**  3.4.5 and 3.5.1

**Search Terms:** Conditional types, Type inference

**Code**

```ts
interface I {a: string}
type Cond<T> = T extends any ? Partial<T>: T;

declare function f<T>(x1: T, x2: Cond<T>): T;
function g(p1: I, p2: Cond<I>): I {
  return f(p1, p2); // error here
}
```

**Expected behavior:**
No errors. A trivial type inference is `T=I`. This is also the behavior of TypeScript 3.3.

**Actual behavior:**
`error TS2322: Type 'Partial<I>' is not assignable to type 'I'.`
From that error message the actual inference is `T=Partial<I>` (or `Cond<I>`), which seems like a regression to me.

**Playground Link:** 
[playground link](https://www.typescriptlang.org/play/#src=interface%20I%20%7Ba%3A%20string%7D%0D%0Atype%20Cond%3CT%3E%20%3D%20T%20extends%20any%20%3F%20Partial%3CT%3E%3A%20T%3B%0D%0A%0D%0Adeclare%20function%20f%3CT%3E(x1%3A%20T%2C%20x2%3A%20Cond%3CT%3E)%3A%20T%3B%0D%0Afunction%20g(p1%3A%20I%2C%20p2%3A%20Cond%3CI%3E)%3A%20I%20%7B%0D%0A%20%20return%20f(p1%2C%20p2)%3B%20%2F%2F%20error%20here%0D%0A%7D)

**Related Issues:** 
There are some issues around conditional types but nothing quite a duplication, I think.
https://github.com/microsoft/TypeScript/issues/31601 requires function overloads
https://github.com/microsoft/TypeScript/issues/24085 is a problem with enums
https://github.com/microsoft/TypeScript/issues/30489 is more about type inference on the conditional type itself, not the generic type when the conditional type is used
https://github.com/microsoft/TypeScript/issues/30341 is similar but it's marked as fixed

@rkirov

**Comment by RyanCavanaugh** [comment]
What is the intent of writing `T extends any` ?

**Comment by bowenni** [comment]
> What is the intent of writing `T extends any` ?

To create a conditional type. I think the problem only happens with a conditional type.

If the example is oversimplified and doesn't make sense, below is another example, without the `T extends any`.

```ts
interface I { a: string }
type DeepPartial<T> =
    T extends object ? {[K in keyof T]?: DeepPartial<T[K]>} : T;

declare function f<T>(t: T, partial: DeepPartial<T>): T;
function g(p1: I, p2: Partial<I>): I {
  return f(p1, p2); // error Type 'Partial<I>' is not assignable to type 'I'.
}
```

**Comment by RyanCavanaugh** [comment]
```ts
interface I { a: string }
type DeepPartial<T> =
    T extends object ? {[K in keyof T]?: DeepPartial<T[K]>} : T;

declare function f<T>(t: T, partial: DeepPartial<T>): T;
function g(p1: I, p2: Partial<I>): I {
    // T inferred to be Partial<I>; should be I
  return f(p1, p2); // error Type 'Partial<I>' is not assignable to type 'I'.
}
```

**Comment by ahejlsberg** [comment]
This is caused by #30010. Part of why we inferred to a union type was to get lower priority candidates for inferences to naked type variables, but that got lost with the fix in #30010. We just need to reinstate that part.

**Comment by weswigham** [comment]
> This is caused by #30010. 

To be fair, at the time I'm pretty sure we had done away with the naked type variable priority; we just didn't think to add it to the conditional inference bit when we added it back. :P
